desktop/main.go

```diff
@@ -1,89 +1,259 @@
-		AstilectronOptions: astilectron.Options{
-			AppName:            AppName,
-			AppIconDarwinPath:  "resources/icon.icns",
-			AppIconDefaultPath: "resources/icon.png",
-			// TODO: Revisit: slightly dangerous because this will ignore _all_ certificate errors
-			ElectronSwitches: []string{"ignore-certificate-errors", "true"},
-		},
+	gateway = fmt.Sprintf("http://%s", textile.GetGatewayAddress())
 
-		Debug:          *debug,
-		Homepage:       "index.html",
-		MessageHandler: handleMessages,
-		OnWait:         start,
-		RestoreAssets:  RestoreAssets,
-		WindowOptions: &astilectron.WindowOptions{
-			BackgroundColor: astilectron.PtrStr("#333333"),
-			Center:          astilectron.PtrBool(true),
-			Height:          astilectron.PtrInt(633),
-			Width:           astilectron.PtrInt(1024),
-		},
-	}); err != nil {
-		astilog.Fatal(errors.Wrap(err, "running bootstrap failed"))
+	// sleep for a bit on the landing screen, it feels better
+	time.Sleep(SleepOnLoad)
+
+	// send cookie info to front-end
+	sendData("login", map[string]interface{}{
+		"name":    "SessionId",
+		"value":   "not used",
+		"gateway": gateway,
+	})
+
+	// check if we're configured yet
+	threads := textile.Wallet.Threads()
+	if len(threads) > 0 {
+		sendPreReady()
+		window.Hide()
+		expandWindow()
+
+		// load threads for UI
+		var threadsJSON []map[string]interface{}
+		for _, thrd := range threads {
+			threadsJSON = append(threadsJSON, map[string]interface{}{
+				"id":   thrd.Id,
+				"name": thrd.Name,
+			})
+		}
+		sendData("ready", map[string]interface{}{
+			"threads": threadsJSON,
+		})
+		window.Show()
+		window.Focus()
+
+	} else {
+		// get qr code for setup
+		qr, pk, err := getQRCode()
+		if err != nil {
+			astilog.Error(err)
+			return err
+		}
+		sendData("setup", map[string]interface{}{
+			"qr": qr,
+			"pk": pk,
+		})
+	}
+
+	return nil
+}
+
+func sendMessage(name string) {
+	window.SendMessage(map[string]string{"name": name})
+}
+
+func sendData(name string, data map[string]interface{}) {
+	data["name"] = name
+	window.SendMessage(data)
+}
+
+func handleMessage(_ *astilectron.Window, m bootstrap.MessageIn) (interface{}, error) {
+	switch m.Name {
+	case "thread.load":
+		var threadId string
+		if err := json.Unmarshal(m.Payload, &threadId); err != nil {
+			return nil, err
+		}
+		html, err := getThreadPhotos(threadId)
+		if err != nil {
+			return nil, err
+		}
+		return map[string]interface{}{
+			"html": html,
+		}, nil
+	default:
+		return map[string]interface{}{}, nil
+	}
+}
+
+func subscribe(thrd *thread.Thread) {
+	for {
+		select {
+		case update, ok := <-thrd.Updates():
+			if !ok {
+				return
+			}
+			sendData("thread.update", map[string]interface{}{
+				"update": update,
+			})
+		}
+	}
+}
+
+func getQRCode() (string, string, error) {
+	// get our own public key
+	pk, err := textile.Wallet.GetPubKeyString()
+	if err != nil {
+		return "", "", err
+	}
+
+	// create a qr code
+	url := fmt.Sprintf("https://www.textile.io/clients?key=%s", pk)
```

Comment:
app requires a confirmation code e.g. 3425. the thought behind that was to show it on the desktop to ensure your phone correctly paired to the right device before sending keys out

Reply:
it'd just be code=3425

Reply:
oh right, forgot i removed that. i think the address serves the same purpose and is more informative / secure... so when the phone goes to confirm the pairing it can just say, "about to pair with this address xxx" (the pub key below the qr code), and the user can double check if they want.

Reply:
perfect

Reply:
we'll also need to let the mobile side name the device, which would presumably happen before the confirmation (?)